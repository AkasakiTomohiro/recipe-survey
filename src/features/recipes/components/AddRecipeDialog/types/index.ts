/**
 * @file index.ts
 * @author Artan's Projects
 * @copyright © 2024 Artan's Projects. All rights reserved.
 */
export interface IAddRecipeDialogProps {

  /**
   * ダイアログの表示・非表示
   */
  open: boolean;

  /**
   * ダイアログのcloseイベント
   */
  onClose: () => void;
}