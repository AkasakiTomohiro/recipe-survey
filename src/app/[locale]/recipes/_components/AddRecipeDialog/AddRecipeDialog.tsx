/**
 * @file AddRecipeDialog.tsx
 * @author Artan's Projects
 * @copyright © 2024 Artan's Projects. All rights reserved.
 */

// import { FormProvider, useForm } from 'react-hook-form';

import { useState } from 'react';

import {
    ControllerCheckboxForm, ControllerSelectForm, ControllerTextForm, IControllerCheckboxFormProps, IControllerSelectFormProps,
    IControllerTextFormProps
} from '@/components/ControllerForm';
import { CloseIcon } from '@/components/Icon';
import { useTransLocations } from '@/i18n';
import { Recipe, useAddRecipe, useSelectedProfileId } from '@/providers/profile/hooks';
import { DialogTitle, Grid, IconButton, Step, StepLabel } from '@mui/material';
import MuiDialog from '@mui/material/Dialog';
import MuiStepper from '@mui/material/Stepper';

import { IAddRecipeDialogProps } from './types';

const TextForm = (props: IControllerTextFormProps<Recipe>): JSX.Element => {
  return (<ControllerTextForm {...props}/>);
};
const SelectForm = (props: IControllerSelectFormProps<Recipe>): JSX.Element => {
  return (<ControllerSelectForm {...props}/>);
};
const CheckboxForm = (props: IControllerCheckboxFormProps<Recipe>): JSX.Element => {
  return (<ControllerCheckboxForm {...props}/>);
};

type IStepperProps = {
  /**
   * ステッパーのタイトル一覧
   */
  steps: string[];

  /**
   * 選択中のステップ
   */
  activeStep: number;

  /**
   * 
   */
  skipped: Set<number>;
}

/**
 * Stepper
 */
const Stepper = (props: IStepperProps): JSX.Element => {
  return (
    <MuiStepper activeStep={props.activeStep} sx={{ marginRight: '32px' }}>
      {props.steps.map((label, index) => {
        const stepProps: { completed?: boolean } = {};
        const labelProps: {
            optional?: React.ReactNode;
          } = {};
        if(props.skipped.has(index)) {
          stepProps.completed = false;
        }
        return (
          <Step key={label} {...stepProps}>
            <StepLabel {...labelProps}>
              {label}
            </StepLabel>
          </Step>
        );
      })}
    </MuiStepper>
  );
}
export const AddRecipeDialog = (props: IAddRecipeDialogProps): JSX.Element => {
  const t = useTransLocations();
  const [activeStep, setActiveStep] = useState(0);
  const [skipped, setSkipped] = useState(new Set<number>());
  const [profileId] = useSelectedProfileId();
  const addRecipe = useAddRecipe(profileId);
  const steps = [t('recipes.registerRecipe.basicInfo'), t('recipes.registerRecipe.materialAndProduct'), t('recipes.registerRecipe.confirm')];
  
  const onClose = (): void => {
    props.onClose();
  };

  const onSubmit = (): void => {
  };

  return (
    <MuiDialog
      open={props.open}
      fullWidth={true}
      maxWidth="md"
      scroll="paper"
      aria-labelledby="select-preset"
      aria-describedby="select-preset"
    >
      <DialogTitle>
        {t('recipes.registerRecipe.title')}
        <Stepper steps={steps} activeStep={activeStep} skipped={skipped} />
        {props.onClose && (
          <IconButton
            aria-label="close"
            onClick={props.onClose}
            sx={{
              position: 'absolute',
              right   : 8,
              top     : 8,
              color   : (theme) => theme.palette.grey[500]
            }}
          >
            <CloseIcon />
          </IconButton>
        )}
      </DialogTitle>
      <>
        {/* <Grid container spacing={3}>
          <TextForm label={t('recipes.headers.recipeName')} name="recipeName" />
          <TextForm label={t('recipes.headers.recipeAlternativeName')} name="recipeAlternativeName" />
          <SelectForm label={t('recipes.headers.machine')} name="machine" list={[]}/>
          <TextForm label={t('recipes.headers.progressTime')} name="progressTime" type="number" />
          <TextForm label={t('recipes.headers.energy')} name="energy" type="number" />
          <SelectForm label={t('recipes.headers.energyUnit')} name="energyUnit" list={[]} />
        </Grid> */}
      </>
    </MuiDialog>
  );
}
