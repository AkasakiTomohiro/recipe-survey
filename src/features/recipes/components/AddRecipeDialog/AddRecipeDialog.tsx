/**
 * @file AddRecipeDialog.tsx
 * @author Artan's Projects
 * @copyright © 2024 Artan's Projects. All rights reserved.
 */

import { useCallback, useState } from 'react';
import { FormProvider, useForm } from 'react-hook-form';
import { useInternationalization } from '~/hooks/UseInternationalization';
import {
    Recipe, RecipeBasicInfo, RecipeBasicInfoValidator, RecipeMaterialAndProduct, RecipeMaterialAndProductValidator, RecipeValidator
} from '~/lib/RecipeValidator';
import { useAddRecipe, useSelectedProfileId } from '~/providers/profile/hooks';
import { zodErrorMaps } from '~/utils/ZodErrorMap';

import { zodResolver } from '@hookform/resolvers/zod';
import { Button, DialogActions, DialogContent, DialogTitle, Step, StepLabel } from '@mui/material';
import MuiDialog from '@mui/material/Dialog';
import MuiStepper from '@mui/material/Stepper';

import { RecipeBasicInfoContents } from './components/RecipeBasicInfo/RecipeBasicInfo';
import { RecipeMaterialAndProductContents } from './components/RecipeMaterialAndProduct/RecipeMaterialAndProduct';
import { IAddRecipeDialogProps } from './types';

type IStepperProps = {
  /**
   * ステッパーのタイトル一覧
   */
  steps: string[];

  /**
   * 選択中のステップ
   */
  activeStep: number;

  /**
   * 
   */
  skipped: Set<number>;
}

/**
 * Stepper
 */
const Stepper = (props: IStepperProps): JSX.Element => {
  return (
    <MuiStepper activeStep={props.activeStep} sx={{ marginRight: '32px' }}>
      {props.steps.map((label, index) => {
        const stepProps: { completed?: boolean } = {};
        const labelProps: {
            optional?: React.ReactNode;
          } = {};
        if(props.skipped.has(index)) {
          stepProps.completed = false;
        }
        return (
          <Step key={label} {...stepProps}>
            <StepLabel {...labelProps}>
              {label}
            </StepLabel>
          </Step>
        );
      })}
    </MuiStepper>
  );
}

export const AddRecipeDialog = (props: IAddRecipeDialogProps): JSX.Element => {
  const { t } = useInternationalization();
  const [activeStep, setActiveStep] = useState(0);
  const [skipped, setSkipped] = useState(new Set<number>());
  const [profileId] = useSelectedProfileId();
  const addRecipe = useAddRecipe(profileId);
  const steps = [t('recipes.registerRecipe.basicInfo'), t('recipes.registerRecipe.materialAndProduct'), t('common.confirm')];

  const productInitialValue = {
    product  : '',
    byProduct: false,
    type     : '',
    amount   : 0
  };

  const materialInitialValue = {
    material: '',
    amount  : 0,
    type    : ''
  };
  const unitList: string[] = ['testUnit'];
  const machineList: string[] = ['製作機', '組立機', '製造機', '精製施設', '混合機', '充填機', '粒子加速器', '製錬炉', '鋳造炉'];

  /**
   * 基本情報ページのバリデータ
   */
  const methodsPage1 = useForm<RecipeBasicInfo>({
    mode         : 'onChange',
    defaultValues: {
      recipeName           : '',
      recipeAlternativeName: '',
      energy               : 0,
      progressTime         : 0,
      machine              : '',
      energyUnit           : ''
    },
    resolver: zodResolver(RecipeBasicInfoValidator, { errorMap: zodErrorMaps })
  });
  const handleSubmitPage1 = useCallback((func: () => Promise<void>) => methodsPage1.handleSubmit(func), [methodsPage1]);

  /**
   * 素材と成果物ページのバリデータ
   */
  const methodsPage2 = useForm<RecipeMaterialAndProduct>({
    mode         : 'onChange',
    defaultValues: {
      materials            : [materialInitialValue],
      products               : [productInitialValue]
    },
    resolver: zodResolver(RecipeMaterialAndProductValidator, { errorMap: zodErrorMaps })
  });
  const handleSubmitPage2 = useCallback((func: () => Promise<void>) => methodsPage2.handleSubmit(func), [methodsPage2]);

  /**
   * 確認ページのバリデータ
   * TODO:
   */
  const methodsPage3 = useForm<Recipe>({
    mode         : 'onChange',
    defaultValues: {
      id                   : 'new_recipeId',
      recipeName           : '',
      recipeAlternativeName: '',
      energy               : 0,
      progressTime         : 0,
      machine              : '',
      energyUnit           : '',
      materials            : [materialInitialValue],
      products             : [productInitialValue]
    },
    resolver: zodResolver(RecipeValidator, { errorMap: zodErrorMaps })
  });
  const handleSubmitPage3 = useCallback((func: () => Promise<void>) => methodsPage3.handleSubmit(func), [methodsPage3]);

  /**
   * activeStepをひとつ次に進める関数
   */
  const handleNext = useCallback(async () => {
    let newSkipped = skipped;
    if(skipped.has(activeStep)) {
      newSkipped = new Set(newSkipped.values());
      newSkipped.delete(activeStep);
    }

    setActiveStep((prevActiveStep) => prevActiveStep + 1);
    setSkipped(newSkipped);
  }, [activeStep, skipped]);

  /**
   * activeStepをひとつ前に戻す関数
   */
  const handleBack = useCallback((): void => {
    if(activeStep === 0) {
      return;
    }
    setActiveStep((prevActiveStep) => prevActiveStep - 1);
  }, [activeStep]);

  /**
   * 完了ボタン押下時の関数
   */
    const handleComplete = useCallback(async () => {
    }, []);

  /**
   * 「次へ」or「完了」ボタン押下時の関数
   */
  const handleSubmit = (() => {
    // if(activeStep === 0) {
    //   return handleSubmitPage1(handleNext);
    // }
    // if(activeStep === 1) {
    //   return handleSubmitPage2(handleNext);
    // }
    // return handleSubmitPage3(handleComplete);
    handleNext();
  });

  return (
    <MuiDialog
      open={props.open}
      fullWidth={true}
      maxWidth="lg"
      scroll="paper"
      aria-labelledby="select-preset"
      aria-describedby="select-preset"
    >
      <DialogTitle>
        {t('recipes.registerRecipe.title')}
      </DialogTitle>
      <DialogTitle>
        <Stepper steps={steps} activeStep={activeStep} skipped={skipped} />
      </DialogTitle>

      <DialogContent>
        {activeStep === 0 && 
          <FormProvider {...methodsPage1}>
            <RecipeBasicInfoContents machineList={machineList} unitList={unitList}/>
          </FormProvider>
        }

        {activeStep === 1 && 
          <FormProvider {...methodsPage1}>
            <RecipeMaterialAndProductContents materialInitialValue={materialInitialValue} productInitialValue={productInitialValue}/>
          </FormProvider>
        }
         
        {activeStep === 2 && <div>{t('common.confirm')}</div>}
      </DialogContent>

      <DialogActions sx={{ justifyContent: activeStep === 0 ? 'end' : 'space-between' }}>
        {activeStep !== 0 && 
          <Button variant="text" color="secondary" onClick={handleBack}>
            {t('common.back')}
          </Button>
        }
        <Button variant="text" color="secondary" onClick={handleSubmit}>
          {activeStep === 0 && t('common.next')}
          {activeStep === 1 && t('common.confirm')}
          {activeStep === 2 && t('common.register')}
        </Button>
      </DialogActions>
    </MuiDialog>
  );
}
