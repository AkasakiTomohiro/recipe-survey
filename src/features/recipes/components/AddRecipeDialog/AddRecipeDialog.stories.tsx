/**
 * @file AddRecipeDialog.stories.tsx
 * @author Artan's Projects
 * @copyright Â© 2024 Artan's Projects. All rights reserved.
 */

import type { Meta, StoryObj } from '@storybook/react';
import { AddRecipeDialog } from './AddRecipeDialog';

const meta: Meta<typeof AddRecipeDialog> = {
  title : 'features/Recipe/AddRecipeDialog',
  render: (args): JSX.Element => <AddRecipeDialog {...args} />
};

export default meta;

type Story = StoryObj<typeof AddRecipeDialog>;

export const Primary: Story = {
  args: {
    open   : true,
    onClose: () => console.log('close')
  }
};
