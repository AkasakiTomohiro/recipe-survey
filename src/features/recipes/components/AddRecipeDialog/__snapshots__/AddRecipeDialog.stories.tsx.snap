// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recipes/components/AddRecipeDialog Primary smoke-test 1`] = `""`;
